aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -28,32 +28,51 @@
 
 #include <map>
 
+namespace at {
+namespace native {
+
 // NOTE:
-// YOU ARE NOT OBLIGED TO USE THESE MACROS
+// YOU ARE NOT OBLIGED TO USE THESE HELPERS
 // If you're writing something more specialized, please don't try to make them
 // work for your case, but just write something new instead.
 
-namespace at {
-namespace native {
+// Helper functions that reduce redundant code in implementing the most typical kind of unary operators. This allows
+// some preprocessing that are unique to some operators (more is forsee-able in the future) and is more flexible and
+// elegant than defining a flat fat macro that implements everything.
+template <typename Stub>
+static inline Tensor& unary_op_out_impl(Tensor& result, const Tensor& self, Stub& stub) {
+  auto iter = TensorIterator::unary_op(result, self,
+    /*check_mem_overlap=*/true);
+  stub(iter.device_type(), iter);
+  return result;
+}
 
-Tensor bitwise_not(const Tensor& self) {
+template <typename OutImpl>
+static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
   Tensor result = at::empty({0}, self.options());
-  return at::bitwise_not_out(result, self);
+  return out_impl(result, self);
 }
 
-Tensor& bitwise_not_(Tensor& self) {
-  return at::bitwise_not_out(self, self);
+template <typename OutImpl>
+static inline Tensor& unary_op_impl_(Tensor& self, OutImpl& out_impl) {
+  return out_impl(self, self);
 }
 
-Tensor& bitwise_not_out(Tensor& result, const Tensor& self) {
-  checkBackend("bitwise_not", result, self.type().backend());
-  auto iter = TensorIterator::unary_op(result, self,
-    /*check_mem_overlap=*/true);
-  bitwise_not_stub(iter.device_type(), iter);
-#ifdef BUILD_NAMEDTENSOR
-  at::namedinference::propagate_names(result, self);
-#endif
-  return result;
+Tensor bitwise_not(const Tensor& self) { return unary_op_impl(self, bitwise_not_out); }
```

Comment:
what is bitwise_not_out?  Is it at::bitwise_not_out?  at::native::bitwise_not_out?